packages/nest/src/oauth/data/datasource/mappers/oauth-client-grant.raw-sql-to-entity.mapper.ts

```diff
@@ -11,11 +11,11 @@ import {
 export class OAuthClientGrantRawSqlToEntityMapper implements Mapper<RawSQLData, OAuthClientGrantEntity> {
     map(from: RawSQLData): OAuthClientGrantEntity {
         return new OAuthClientGrantEntity(
-            from[OAuthColumnId],
-            from[OAuthColumnCreatedAt],
-            from[OAuthColumnUpdatedAt],
-            from[OAuthClientGrantColumnGrantName],
-            from[OAuthClientGrantColumnClientId],
+            from[OAuthColumnId] as number,
+            from[OAuthColumnCreatedAt] as Date,
+            from[OAuthColumnUpdatedAt] as Date,
```

Comment:
TypeORM already returns the correct type instances (e.g. date columns are already `Date` instances), the thing is that we get a `Record<string, unknown>` from `RawSQLDataSource`. We could change this to `Record<string, any>` and we could skip the explicit type casting. Not sure which is the correct thing to do in this case.

Related to https://github.com/mobilejazz/harmony-typescript/pull/76#discussion_r984810983.

cc @lucianosantana

Reply:
@orioljp @lucianosantana ping

Reply:
Uhmm... the situation is worse than I imagined. See the screenshot below.

Columns which don't have value (`NULL` in the DB) arrive as `null`, **these probably should be converted to `undefined`**. The type of the columns which have values are correct… except for `updated_at`, that in the DB has value `0000-00-00 00:00:00.000000` and the `RawSQLDataSource` returns that same value (hence `string` instead of `Date`).

This is something we need to think about; the mapper is actually too optimistic (well, it's not doing any work…). Maybe we need to store a definition of the table (which are the fields, and of which type); then maybe we can have "pre-mapper" in `RawSQLDataSource` that takes care of these cases. Or maybe wrap the answer from TypeORM into a new `Row` class which exposes some getters (which at the same time does the extra mapping). Example:

```
row.get('updated_at') // converts `null` to `undefined`
```

If we can't guarantee the types when converting from the DB to our entities this is going to bit us at any moment. No strict-mode is going to catch any of this. Garbabe in, garbage out. **So, answering one of my questions, I think we should treat this data as `unknown`, and force ourselves to type-check these at runtime before mapping.** Is important to get this boundary right.

**That being said, I think we should solve this in another issue/PR.**

---

![image](https://user-images.githubusercontent.com/188612/193994073-5cba71e8-6a34-4669-b6a6-2e68d5b528a5.png)

Reply:
This is too big of a melon 🍈 to handle in this PR, I propose we move the conversation to this issue: https://github.com/mobilejazz/harmony-typescript/issues/106

Reply:
I agree @doup , let's work on this in a new PR. @Alex-DA can help showing how Harmony-PHP works.